### PR TITLE
PHP: allow xDS interop client to start RPCs asynchronously

### PR DESCRIPTION
**Pull request by stanley-cheung**
This PR allows the `timeout` xDS interop test case to pass for PHP.

Summary:
- This PR allows a new way for PHP to start each RPC in a separate process, to allow us to pass some additional test case like `timeout`.
- There's an error `"ADS call status
received. Status = 8, details = 'Quota exceeded for quota metric 'xDS new stream requests' and limit 'xDS new
stream requests per minute' of service 'trafficdirector.googleapis.com' for consumer 'project_number:830293263384'.',` but only at high qps. So this seems to suggest that this approach (starting each RPC in a separate process) is not sustainable at high qps in this gcp project setup. 
- I think this approach may still be useful for cases like `timeout`, where we just need a way to start RPCs asynchronously and is OK with a lower QPS for PHP.

Details:
- Instead of asking the `run_xds_tests.py` runner to start the PHP client as `php xds_client.php ...`, we will be launching the PHP client as a shell script first `./src/php/bin/run_xds_client.sh ...`.
- This shell script is responsible for launching the existing PHP interop client `xds_client.php`, which will continue to respond to stats request from the `run_xds_tests.py`.
- The shell script will now launch a new python script `src/php/bin/xds_manager.py`. This py script will continually check whether the main `xds_client.php` will need any RPCs (`UnaryCall`, `EmptyCall`) to be executed / started in an asynchronous manner (i.e. most likely due to the need to honor those `rpc-behavior` metadata configured by the `run_xds_tests.py` runner). 
- This new python script will start a new child PHP process to execute each of those RPCs. This is the only way we can have multiple concurrent PHP RPCs opened. The script will regularly check whether any of those outstanding child PHP processes has finished, and report back to the main `xds_client.php` what each of the child RPCs' exit status code is.
- Two temp files will be needed for these communications to work.
   - `tmp_file1` will be written by `xds_client.php` when there's a need to start executing RPCs in these new child PHP processes. It will be read by `xds_manager.py` and then start more PHP child processes.
   - `tmp_file2` will be written by `xds_manager.py` when a PHP child process finishes. It will be read by `xds_client.php` and the result aggregated and reported back to the `run_xds_tests.py` test runner.
   -  Some basic `flock` and `ftruncate` have been implemented.
- Implemented the new `XdsUpdateClientConfigureService::Configure()` and `LoadBalancerStatsService::getClientAccumulatedStats()` RPCs as required by the `run_xds_tests.py` test spec.
- Similar to the approach with Ruby, we have abstracted out all the knobs and switches that can affect client RPCs execution into an `RpcConfig` object.

Note that this approach will not work for `circuit_breaking` as concurrent in-flight RPCs are only counted local to a process. As we are starting a child PHP process for each RPC here, these counters will only ever be run up to 1.

Outstanding issue:
- Some test flakiness is being observed with the `timeout` test case. I can consistently pass the `timeout` test with `qps=25`. Anything higher than that, I sometimes get a big stream of the following error, which is causing all the client RPCs to fail. But this only happens sometimes and I could not consistently pass `timeout` with the default `qps=100`. Will need to look into this error. So as of now, I am testing `timeout` for PHP with only `qps=25`.
```
I0312 21:24:34.142576919   20565 xds_client.cc:814]          [xds_client 0x562c59d3a470] sending ADS request:
 type=type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment version= nonce= error="No Error"
resources=cloud-internal-istio:cloud_mp_830293263384_3441680000408249382                                         
I0312 21:24:34.143573625   20619 xds_client.cc:1247]         [xds_client 0x557b55e62470] ADS call status
received. Status = 8, details = 'Quota exceeded for quota metric 'xDS new stream requests' and limit 'xDS new
stream requests per minute' of service 'trafficdirector.googleapis.com' for consumer 'project_number:830293263384'.',
(chand: 0x557b55e5b3d0, ads_calld: 0x557b55c9c9f0, call: 0x557b55f9c1e0), error '"No Error"'
I0312 21:24:34.143618657   20619 xds_client.cc:630]          [xds_client 0x557b55e62470] Failed to connect to
xds server (chand: 0x557b55e5b3d0) retry timer will fire in 1000ms.                                          
E0312 21:24:34.143657648   20619 xds_resolver.cc:761]        [xds_resolver 0x557b55da1f40] received error from
XdsClient: {"created":"@1615584274.143635116","description":"xds call failed","file":"src/core/ext/xds/xds_client.cc",
"file_line":1261} 
```
-  With this error, looks like it's not a sustainable approach to start each RPC in a separate process. But with a low qps, at least we can get PHP to pass some additional test scenarios which involve starting RPCs asynchronously.

Ad-hoc run of all PHP xDS interop test (all passed): https://fusion2.corp.google.com/invocations/a39deaf3-91b8-4274-bb5e-f381dec35f44/targets

### CODE REVIEW

**Comment by stanley-cheung** [comment]
@dfawley Thanks in advance for the review. You probably don't need to review the PHP code, but I'd like to get your approval on this general approach of starting each PHP RPC in a separate process. With the quota error I mentioned in the PR body, this doesn't look like a generally sustainable approach (looks like all the xDS set up is rather centric to everything happening within the same process). But with this approach, at least we can have an option of giving PHP a chance to start multiple RPCs asynchronously. 

I have all the existing tests to use the old approach (everything happening in a blocking way in the same process) and only `timeout` to use this new approach. Going forward this should be only used unless it's absolutely necessary.

Let me know what you think! Thanks.

File: src/php/bin/xds_manager.py
Comment (dfawley) [inline src/php/bin/xds_manager.py:40]: Why are these set explicitly, instead of picking them up from the environment?

**Review by dfawley — APPROVED** [review]
The approach makes sense and seems fine.  I'd be okay with an even lower QPS, like 5-10, if it helps the stability of the test.  Also, I'm a little concerned about using this approach for only this one test case.  It could cover up issues that running with the others could potentially expose.

Does the php client support streaming RPCs, and would it support multiple streams on one channel?  If so, we could think about rewriting these tests to use streaming RPCs instead.

**Comment by stanley-cheung** [comment]
I left this for a few days, came back to this, rebased to the latest `master` and tried to make some adjustments based on @dfawley 's review comments, and now everything breaks. Debugged the heck out of this and couldn't figure out why my RPCs are failing. Finally decided to "un"-rebase and get back to my old state and everything works immediately. So something in the last 50 commits in the `master` branch does not react well with this.....

**Comment by stanley-cheung** [comment]
> The approach makes sense and seems fine. I'd be okay with an even lower QPS, like 5-10, if it helps the stability of the test.

Comment addressed. I am setting the qps for the `timeout` test to 20 for now.

> Also, I'm a little concerned about using this approach for only this one test case. It could cover up issues that running with the others could potentially expose.

Comment addressed. Now all PHP xds test cases are going through the new approach.

> 
> Does the php client support streaming RPCs, and would it support multiple streams on one channel? If so, we could think about rewriting these tests to use streaming RPCs instead.

We can probably address in the future, but here's what PHP can do - we can start multiple bidi-streams in the same channel. But any call to `->read()` and `->write()` on a PHP client _is_ blocking. So if we do go this route in the future, we have to make sure the server regularly write some no-op messages to unblock any `read()` from the client side.

File: src/php/bin/xds_manager.py
Comment (stanley-cheung) [reply src/php/bin/xds_manager.py]: Comment addressed. Removed the hardcoded env var.

**Comment by stanley-cheung** [comment]
So the PHP + v3 test failure was finally figured out. See issue #25926. We are finally back on track to finally merge this PHP + `timeout` test, and I am finally able to rebase this branch to the latest `master`. 

Going to run the full suite a test one more time just to be safe.

**Comment by stanley-cheung** [comment]
All v3 test cases passed: https://fusion2.corp.google.com/invocations/d59e4389-c13f-443b-b4b7-5ba4d81de019/targets

**Pull request merged**
